zio-http/src/main/scala/zio/http/endpoint/openapi/JsonSchema.scala

```diff
@@ -0,0 +1,711 @@
+      case TextCodec.StringCodec      => JsonSchema.String
+      case TextCodec.IntCodec         => JsonSchema.Integer(JsonSchema.IntegerFormat.Int32)
+      case TextCodec.LongCodec        => JsonSchema.Integer(JsonSchema.IntegerFormat.Int64)
+      case TextCodec.BooleanCodec     => JsonSchema.Boolean
+      case TextCodec.UUIDCodec        => JsonSchema.String
+    }
+
+  def fromSegmentCodec(codec: SegmentCodec[_]): JsonSchema =
+    codec match {
+      case SegmentCodec.BoolSeg(_)                    => JsonSchema.Boolean
+      case SegmentCodec.IntSeg(_)                     => JsonSchema.Integer(JsonSchema.IntegerFormat.Int32)
+      case SegmentCodec.LongSeg(_)                    => JsonSchema.Integer(JsonSchema.IntegerFormat.Int64)
+      case SegmentCodec.Text(_)                       => JsonSchema.String
+      case SegmentCodec.UUID(_)                       => JsonSchema.String
+      case SegmentCodec.Annotated(codec, annotations) =>
+        fromSegmentCodec(codec).description(segmentDoc(annotations)).examples(segmentExamples(codec, annotations))
+      case SegmentCodec.Literal(_) => throw new IllegalArgumentException("Literal segment is not supported.")
+      case SegmentCodec.Empty      => throw new IllegalArgumentException("Empty segment is not supported.")
+      case SegmentCodec.Trailing   => throw new IllegalArgumentException("Trailing segment is not supported.")
+    }
+
+  private def segmentDoc(annotations: Chunk[SegmentCodec.MetaData[_]]) =
+    annotations.collect { case SegmentCodec.MetaData.Documented(doc) => doc }.reduceOption(_ + _).map(_.toCommonMark)
+
+  private def segmentExamples(codec: SegmentCodec[_], annotations: Chunk[SegmentCodec.MetaData[_]]) =
+    Chunk.fromIterable(
+      annotations.collect { case SegmentCodec.MetaData.Examples(example) => example.values }.flatten.map { value =>
+        codec match {
+          case SegmentCodec.Empty           => throw new IllegalArgumentException("Empty segment is not supported.")
+          case SegmentCodec.Literal(_)      => throw new IllegalArgumentException("Literal segment is not supported.")
+          case SegmentCodec.BoolSeg(_)      => Json.Bool(value.asInstanceOf[Boolean])
+          case SegmentCodec.IntSeg(_)       => Json.Num(value.asInstanceOf[Int])
+          case SegmentCodec.LongSeg(_)      => Json.Num(value.asInstanceOf[Long])
+          case SegmentCodec.Text(_)         => Json.Str(value.asInstanceOf[String])
+          case SegmentCodec.UUID(_)         => Json.Str(value.asInstanceOf[java.util.UUID].toString)
+          case SegmentCodec.Trailing        =>
+            throw new IllegalArgumentException("Trailing segment is not supported.")
+          case SegmentCodec.Annotated(_, _) =>
+            throw new IllegalStateException("Annotated SegmentCodec should never be nested.")
+        }
+      },
+    )
+
+  def fromZSchema(schema: Schema[_], refType: SchemaStyle = SchemaStyle.Inline): JsonSchema =
```

Comment:
IMO `fromZSchema` should be able to accumulate component schemas. For example `oneOf`, `allOf` cases should be named. If they are not named, this prohibits proper use of discriminators. Also schemas that are used in multiple places should not be expanded multiple times. 

One possible approach would be to allow `fromZSchema` to return multiple  `JsonSchemas`, but I'm sure there are other approaches.

Reply:
The case of expanding multiple times was intentionally ignored. As it does not change meaning, but is just an optimization. And also right now, we build schemas per endpoint and then accumulate. In this case, we can't be very efficient, since we don't know all Endpoints. Also I am not sure how efficient this needs to be, since I don't expect that building OpenAPI is happening more than once per application.
Actually, we could probably generate it at compile time 🤔 
Your argument with the naming I don't get. Could you please give me more details? Or en example?

Reply:
I am referring to this section https://swagger.io/specification/#discriminator-object, where I am not sure the example for `oneOf` with discriminator can be modelled without naming cases. One can probably get pretty far without, but I have seen it in the past.

My concern with duplication is not about efficiency, but ergonomics. Usually the oeverall openapi spec is not for a single endpoint, but for a more complex API, where lots of types appear in multiple places. A human reader of the API will need to read a very long documents with lots of repetitions. A client that is generated from the spec will have basically the same datatype repeated lots of times under different names. If everything was expanded, such types could be generated exactly once.

Reply:
I think I get what you mean. You want the schemas not to be inlined, but part of components, so you can use a discriminator to reference the schema. I'll impl. this

Reply:
Exactly. Wonderful. :)

Reply:
@runtologist Thanks for the helpful review. Maybe you could check the changes 🙂

Reply:
> Enums are still inlined instead of referenced
> 
> Element types of arrays are still inlined instead of referenced.

In the test I see something else. Maybe  matter of what SchemaStyle is used. Do you have an example?

> AFAIU it should just be OriginalType.

I am not so sure about that. I'll try to read specs again and think about what implications that has. Anyhow I do not see this as a show stopper, since it is not wrong. Option[Value] in json is null | Value.

> After the change caseNames on enums are not applied any more. All enum values are the name of the class/object.

I'll check this